File: web/backend/app/routes/followups.py
from fastapi import APIRouter, Depends

from ..deps import require_patient
from ..schemas.followup import (
    FollowUpBookingRequest,
    FollowUpBookingResponse,
    FollowUpDetailResponse,
    FollowUpListResponse,
    FollowUpPaymentActionResponse,
    FollowUpPaymentCodeRequest,
    FollowUpPaymentInitializeRequest,
    FollowUpPaymentInitializeResponse,
    FollowUpPaymentVerifyResponse,
)
from ..services.followup_app_service import (
    create_patient_followup_booking,
    get_patient_followup,
    initialize_followup_payment,
    list_patient_followups,
    mark_followup_pay_later,
    redeem_followup_payment_code,
    verify_followup_payment,
)


router = APIRouter()


@router.get("/upcoming", response_model=FollowUpListResponse)
def upcoming_followups(session: dict = Depends(require_patient)):
    return list_patient_followups(session["user_id"])


@router.post("/book", response_model=FollowUpBookingResponse)
def book_followup(payload: FollowUpBookingRequest, session: dict = Depends(require_patient)):
    return create_patient_followup_booking(session["user_id"], payload.model_dump())


@router.get("/{reference}", response_model=FollowUpDetailResponse)
def followup_detail(reference: str, session: dict = Depends(require_patient)):
    return get_patient_followup(reference, session["user_id"])


@router.post("/{reference}/payment/initialize", response_model=FollowUpPaymentInitializeResponse)
async def initialize_followup_payment_route(
    reference: str,
    payload: FollowUpPaymentInitializeRequest,
    session: dict = Depends(require_patient),
):
    return await initialize_followup_payment(reference, session["user_id"], payload.model_dump())


@router.post("/{reference}/payment/pay-later", response_model=FollowUpPaymentActionResponse)
def followup_pay_later(reference: str, session: dict = Depends(require_patient)):
    return mark_followup_pay_later(reference, session["user_id"])


@router.post("/{reference}/payment/redeem", response_model=FollowUpPaymentActionResponse)
def followup_redeem_payment_code(
    reference: str,
    payload: FollowUpPaymentCodeRequest,
    session: dict = Depends(require_patient),
):
    return redeem_followup_payment_code(reference, session["user_id"], payload.payment_code)


@router.post("/{reference}/payment/verify/{payment_reference}", response_model=FollowUpPaymentVerifyResponse)
async def followup_verify_payment(
    reference: str,
    payment_reference: str,
    session: dict = Depends(require_patient),
):
    return await verify_followup_payment(reference, session["user_id"], payment_reference)

File: web/frontend/src/api/followups.js
import { authHeaders } from "./auth.js";

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://127.0.0.1:8000";

async function parseResponse(response) {
  const body = await response.json();
  if (!response.ok) {
    throw new Error(body?.detail || `Request failed: ${response.status}`);
  }
  return body;
}

export async function fetchPatientFollowups() {
  const response = await fetch(`${API_BASE_URL}/followups/upcoming`, {
    headers: {
      ...authHeaders(),
    },
  });
  return parseResponse(response);
}

export async function fetchFollowup(reference) {
  const response = await fetch(`${API_BASE_URL}/followups/${encodeURIComponent(reference)}`, {
    headers: {
      ...authHeaders(),
    },
  });
  return parseResponse(response);
}

export async function bookFollowup(payload) {
  const response = await fetch(`${API_BASE_URL}/followups/book`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...authHeaders(),
    },
    body: JSON.stringify(payload),
  });
  return parseResponse(response);
}

export async function initializeFollowupPayment(reference, payload = {}) {
  const response = await fetch(
    `${API_BASE_URL}/followups/${encodeURIComponent(reference)}/payment/initialize`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...authHeaders(),
      },
      body: JSON.stringify(payload),
    },
  );
  return parseResponse(response);
}

export async function verifyFollowupPayment(reference, paymentReference) {
  const response = await fetch(
    `${API_BASE_URL}/followups/${encodeURIComponent(reference)}/payment/verify/${encodeURIComponent(paymentReference)}`,
    {
      method: "POST",
      headers: {
        ...authHeaders(),
      },
    },
  );
  return parseResponse(response);
}

export async function markFollowupPayLater(reference) {
  const response = await fetch(
    `${API_BASE_URL}/followups/${encodeURIComponent(reference)}/payment/pay-later`,
    {
      method: "POST",
      headers: {
        ...authHeaders(),
      },
    },
  );
  return parseResponse(response);
}

export async function redeemFollowupPaymentCode(reference, paymentCode) {
  const response = await fetch(
    `${API_BASE_URL}/followups/${encodeURIComponent(reference)}/payment/redeem`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...authHeaders(),
      },
      body: JSON.stringify({ payment_code: paymentCode }),
    },
  );
  return parseResponse(response);
}
